subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/ParallelDownloadsIntegrationTest.groovy

```diff
@@ -206,6 +206,24 @@ class ParallelDownloadsIntegrationTest extends AbstractHttpDependencyResolutionT
             
             configurations { compile }
             def lock = new java.util.concurrent.locks.ReentrantLock()
+
+            class LockingRule implements ComponentMetadataRule {
+
+                java.util.concurrent.locks.Lock lock
+
+                public LockingRule(java.util.concurrent.locks.Lock lock) {
+                    this.lock = lock
+                }
+
+                public void execute(ComponentMetadataContext context) {
```

Comment:
:x: well, this test shows precisely what is wrong. The very goal of class based rules is to isolate them and make sure they can run in parallel. If they share state, we're doing it wrong. So this test only makes sense for closure based rules, and should _FAIL_ with class based rules.